src/Testcontainers.MongoDb/MongoDbBuilder.cs

```diff
@@ -69,6 +76,22 @@ public override MongoDbContainer Build()
         // provided, the log message "Waiting for connections" appears twice.
         // If the user does not provide a custom waiting strategy, append the default MongoDb waiting strategy.
         var mongoDbBuilder = DockerResourceConfiguration.WaitStrategies.Count() > 1 ? this : WithWaitStrategy(Wait.ForUnixContainer().AddCustomWaitStrategy(new WaitUntil(DockerResourceConfiguration)));
+
+        if (!string.IsNullOrEmpty(DockerResourceConfiguration.ReplicaSetName))
+        {
+            mongoDbBuilder = mongoDbBuilder
+                .WithCommand(DockerResourceConfiguration.Command.Concat(["--replSet", DockerResourceConfiguration.ReplicaSetName, "--keyFile", "/tmp/keyfile", "--bind_ip_all"]).ToArray())
+                .WithResourceMapping(Encoding.Default.GetBytes("""
+                    #!/bin/bash
+                    openssl rand -base64 32 > "/tmp/keyfile"
+                    chmod 600 /tmp/keyfile
+                    chown 999:999 /tmp/keyfile
+                    """.Replace("\r", "")), "/docker-entrypoint-initdb.d/01-init-keyfile.sh", UnixFileModes.OtherRead | UnixFileModes.OtherExecute)
+                .WithWaitStrategy(Wait.ForUnixContainer().UntilCommandIsCompleted(
+                    "mongosh -u $MONGO_INITDB_ROOT_USERNAME -p $MONGO_INITDB_ROOT_PASSWORD --quiet --eval " +
+                    $"\"try {{ rs.status().ok }} catch (e) {{ rs.initiate({{'_id':'{DockerResourceConfiguration.ReplicaSetName}', members: [{{'_id':1, 'host':'127.0.0.1:27017'}}]}}).ok }}\""));
```

Comment:
I didn't want to put them into `WithReplicaSet(string)`, since they're changing some other properties, like the command, to add additional parameters. If it was set in that method, and subsequent calls to change them would break the replica set setup, but it wouldn't be obvious why. Setting them here means this should correctly configure the replica set, regardless of other config values that are set.

i.e. calling the following should still work correctly
```
new MongoDbBuilder().WithReplicaSet().WithCommand("--some-flag")
```

Reply:
Is there any benefit to moving the replica init call into the startup callback instead of a wait strategy? In either case, this should run before the container is marked as ready, but after the container has started, and the wait strategy seemed to match the exact requirements for what we do - finish the configuration of the mongodb server before we allow connections in. Happy to change it if the startup callback makes more sense

Reply:
> I didn't want to put them into `WithReplicaSet(string)`, since they're changing some other properties, like the command, to add additional parameters. [...] Setting them here means this should correctly configure the replica set, regardless of other config values that are set.

This is a general issue. Modules come pre-configured and are opinionated. Developers can always modify the builder configuration in ways that break the setup. It's a balancing act between providing enough flexibility and ensuring reliability.

I believe no other module contains this kind of logic within the `Build()` method. Until now, we've only configured and selected the appropriate wait strategy in that method IIRC.

I would like to [suggest](https://github.com/testcontainers/testcontainers-dotnet/commit/a33964aabbd39a13583338ff5334938c397b8fd7) a few changes to keep this consistent with the other modules. Honestly, I prefer moving the `else` block from 
`Build()` to `WithReplicaSet(string)` too, and only keep the adjustment of the wait strategy there, as we do in [other modules](https://github.com/testcontainers/testcontainers-dotnet/blob/87184e3e53b270ac352cdf59af6c2334488c7fd1/src/Testcontainers.Couchbase/CouchbaseBuilder.cs#L85). WDYT?

The PR looks good so far. Thanks a lot for your contribution already.

Reply:
Hey Andre, thanks for your comments and suggestions!
I've taken again at the library internals, and realised I misunderstood how `WithCommand(string[])` works, which is why I took my initial approach.

I moved most of the RS config to `WithReplicaSet(string)`, as you initially suggested.

The only thing I've left is the init wait strategy, and I left it in `Build()`, similar to the other module you linked. I also added it at the end of it, after the other strategies are added, since we need to make sure this strategy runs, even if other custom strategies are added, otherwise the replica set won't be initialised.

Hopefully this looks much better than the initial version. Let me know what you think